[mongoose] Use mongodb.FilterQuery type

avaly opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint mongoose` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: The mongodb types already use this filter type: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mongodb/index.d.ts#L974
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

amcasey commented on types/mongoose/index.d.ts:
```diff
@@ -3590,4 +3591,31 @@ declare module "mongoose" {
   }
+
+  /* Helper type to extract a definition type from a Document type */
+  export type DocumentDefinition<T> = T extends Document & infer D ? D : never;
```
Same comment as in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42652.

avaly replied:
Refactored this implementation with an `Omit` type.

amcasey commented on types/mongoose/index.d.ts:
```diff
@@ -3590,4 +3591,31 @@ declare module "mongoose" {
+   * which would result in a passing conditional, because `never` is included in all types.
+   */
+  export type MongooseFilterQuery<T> = {
+    [P in keyof T]?: P extends '_id' ?
```
If `T` is expected to have a lot of properties, this could be slow, which could affect editor performance.

avaly replied:
This syntax is used in other places inside `mongoose` types definition itself, but also in other popular libraries.

amcasey replied:
In general, it's fine.  However, if you expect `T` to have hundreds of properties, users may notice a degraded editor experience.  It sounds like you don't expect that to be the case here.

amcasey commented:
@avaly Looks like you might have a merge conflict with your other change. :smile:

avaly commented:
I've rebased this PR on top of the latest `master`, and I also incorporated changes from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/43039 here.

amcasey commented:
@avaly How do you want to handle #43039?  Just close it and merge this instead?

avaly commented:
@amcasey I'd rather merge #43039 on its own first, in order to keep the git history clean. Then I'll rebase this PR on top of master again.

amcasey commented:
Sounds good to me.  I'd like to wait for a clean CI though, if that's alright (I realize this broke at least one consumer).

sandersn commented:
The build break from master is fixed, so I restarted Travis.

amcasey commented:
@avaly Time to rebase?

avaly commented:
@amcasey I've rebased this PR. PTAL Thanks!

reflash commented:
My code is not compiling after this PR. Code itself works fine, it’s only the types that are failing. In this case user is a ref document, so it asks me to pass user type.
```js
public async getSubscriptionByUserId(userId: string) {
    return SubscriptionModel.findOne({ user: userId, ... })
        .populate('user')
        .sort({ _id: -1 });
}
```

Here's my schema and model:
```js
export const subscriptionSchema = new mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
    },
    ...
});
```

```js
interface ISubscription {
    user: User;
    ...
}
```

Is there something I'm doing wrong? I had to go back to previous @types/mongoose version to make it compilable again

avaly commented:
@reflash define you interface as:

```
interface ISubscription {
    user: ObjectId | User;
    ...
}
```

pwrnrd commented:
Following up on the conversation started in #42652, I would love to follow up `find` type definition. I'm trying to create a query paginator. Unfortunatly I get an error updating to the latest `@types/mongoose` version. I have a hard time finding out how to type this correctly.

I have the following (slimmed down) code:

```typescript
import mongoose, { Document } from 'mongoose';

...
interface IPaginatedQuery {
	filter?: { [key: string]: any };
	sort?: string;
	last?: number;
	before?: string;
	first?: number;
	after?: string;
}

interface IPaginatedQueryMongoDB extends IPaginatedQuery {
	projection?: string | { [key: string]: any };
}

function buildFindPart<T extends Document>(query: IPaginatedQueryMongoDB): 

mongoose.MongooseFilterQuery<T> {
	const { filter, sort, before, after } = query;

	const queryObject = {
		...filter,
		// ... and other stuff
	};
	return queryObject; // ERROR 
}

function paginatedQuery<T extends Document>(
	query: IPaginatedQueryMongoDB,
	model: mongoose.Model<T>,
): Promise<IPaginatedResults<T> | undefined> {

	const findPart = buildFindPart<T>(query);

	const results = await model
		.find(findPart)
		.lean();
}

````

The error is as follows: 

````
Type '{ [x: string]: any; }' is not assignable to type 'MongooseFilterQuery<T>'.
  Type '{ [x: string]: any; }' is not assignable to type '{ [P in keyof T]?: (P extends "_id" ? Condition<string | ObjectId | { _id: ObjectId; }> : [Extract<T[P], ObjectId>] extends [never] ? Condition<T[P]> : Condition<...>) | undefined; }'
`````

P.S. A little off-topic but would it be a good idea to apply semantic versioning to the type definitions or mention in the README that we're not following semver? I can create a PR to update the README if this is of interest.

avaly commented:
@pwrnrd All `.find()` operations must now be properly typed.

In your specific case, TS is complaining about `{ [x: string]: any; }` which is the type for `query.filter`. You're using this value (spread) as the `queryObject`, which should match `MongooseFilterQuery<T>`. But your filter is defined as `filter?: { [key: string]: any }` inside `IPaginatedQuery`. You'll need to make `IPaginatedQuery` a generic type:

```ts
interface IPaginatedQuery<T extends Document> {
  filter?: mongoose.MongooseFilterQuery<T>
  ...
}
```
